proxy/src/telemetry/metrics/prometheus.rs

```diff
@@ -0,0 +1,521 @@
+use std::default::Default;
+use std::{fmt, u32};
+use std::hash::{Hash, Hasher};
+use std::iter::{self, IntoIterator, Iterator};
+use std::ops;
+use std::sync::{Arc, Mutex};
+
+use futures::future::{self, FutureResult};
+use http;
+use hyper;
+use hyper::header::{ContentLength, ContentType};
+use hyper::StatusCode;
+use hyper::server::{
+    Service as HyperService,
+    Request as HyperRequest,
+    Response as HyperResponse
+};
+use indexmap::{IndexMap};
+
+use ctx;
+use telemetry::event::Event;
+use super::latency::{BUCKET_BOUNDS, Histogram};
+
+/// A set of Prometheus label pairs.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct Labels {
+    inner: Arc<LabelsInner>
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Default)]
+struct LabelsInner {
+    labels: IndexMap<&'static str, String>,
```

Comment:
It looks like `Labels` is used a key to index into an `IndexMap`, but it also contains a series of `IndexMap`s. Just wondering here, do we need the `IndexMap` here inside `Labels`? Seeing as the keys are a `&'static str`, that tells me we already know all the keys at compile time. Could `Labels` instead just be a struct with each key as a field, and an `Option<T>` as the value?

Reply:
This is probably a good call. I considered doing something like that initially, used `IndexMap` for quicker prototyping, and never revisited it. Probably should.